skate-plugin/project-gen/src/jvmMain/kotlin/slack/tooling/projectgen/circuitgen/CircuitComponent.kt

```diff
@@ -0,0 +1,408 @@
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.asTypeName
+import java.io.File
+import java.io.StringWriter
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.io.path.pathString
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.CIRCUIT_INJECT
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.CIRCUIT_UI_EVENT
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.CIRCUIT_UI_STATE
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.COMPOSABLE
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.IMMUTABLE
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.JAVA_INJECT
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.MODIFIER
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.MUTABLE_STATE_FLOW
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.NAVIGATOR
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.PARCELIZE
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.PRESENTER
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.SCREEN_INTERFACE
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.SLACK_DISPATCHER
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.STATE_FLOW
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.UDF_VIEW_MODEL
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.VIEW_MODEL
+import slack.tooling.projectgen.circuitgen.CircuitGenClassNames.Companion.VIEW_MODEL_KEY
+
+interface CircuitComponent {
+  val fileSuffix: String
+
+  fun screenClassName(featureName: String): String = "${featureName.replace(fileSuffix, "")}Screen"
+
+  fun generate(packageName: String, className: String): FileSpec
+
+  fun isTestComponent(): Boolean = false
+
+  fun writeToFile(directory: Path, packageName: String?, className: String) {
+    val baseDirectory =
+      if (isTestComponent()) Path.of(directory.pathString.replace("src/main", "src/test"))
+      else directory
+    val fileSpec = generate(packageName.orEmpty(), className)
+    Files.createDirectories(baseDirectory)
+    val stringWriter = StringWriter()
+    fileSpec.writeTo(stringWriter)
+    val generatedCode = stringWriter.toString().replace("public ", "")
+    File("${baseDirectory}/${className}${fileSuffix}.kt").writeText(generatedCode)
+  }
+}
+
+class CircuitScreen : CircuitComponent {
+  override val fileSuffix = "Screen"
+
+  override fun generate(packageName: String, className: String): FileSpec {
+    /**
+     * Generate Circuit Screen class, eg:
+     *
+     * @Parcelize class ${NAME}Screen : Screen { data class State( val eventSink: Event.() -> Unit =
+     *   {}, ) : CircuitUiState
+     * @Immutable sealed interface Event : CircuitUiEvent }
```

Comment:
Let's put this in code blocks so they're not reformatted. Same for other cases

```suggestion
     * ```kotlin
     * @Parcelize class ${NAME}Screen : Screen { data class State( val eventSink: Event.() -> Unit =
     *   {}, ) : CircuitUiState
     * @Immutable sealed interface Event : CircuitUiEvent }
     * ```
```

Reply:
updated in https://github.com/slackhq/slack-gradle-plugin/pull/799/commits/790aff0b03a4d5bc0e12a2ed4ee3f6cf5dad71a3